packages/slate-react/src/components/text.js

```diff
@@ -95,10 +90,19 @@ class Text extends React.Component {
     const { props } = this
     this.debug('render', { props })
 
-    const { node, schema, state } = props
+    const { decorations, node, state } = props
     const { document } = state
-    const decorators = schema.hasDecorators ? document.getDescendantDecorators(node.key, schema) : []
-    const ranges = node.getRanges(decorators)
+    const { key } = node
+
+    const decs = decorations.filter((d) => {
+      const { startKey, endKey } = d
+      if (startKey == key || endKey == key) return true
+      const startsBefore = document.areDescendantsSorted(startKey, key)
```

Comment:
I agree though, this can be improved.

One improvement I think would be a method that checks `(before, key, after)` or something about checking whether a node "overlaps" a range. Because here we have to do two separate checks which each iterate keys, etc. when really it could be handled in one go.